Storage/SQL/SQLiteHistoryFavicons.swift

```diff
@@ -0,0 +1,439 @@
+            return db.withConnection { conn -> Int in
+                // Blind! We don't see failure here.
+                let id = self.favicons.insertOrUpdateFaviconInTransaction(icon, conn: conn)
+
+                // Now set up the mapping.
+                try conn.executeChange(query, withArgs: args)
+
+                // Try to update the favicon ID column in each bookmarks table. There can be
+                // multiple bookmarks with a particular URI, and a mirror bookmark can be
+                // locally changed, so either or both of these statements can update multiple rows.
+                if let id = id {
+                    icon.id = id
+
+                    try? conn.executeChange("UPDATE bookmarksLocal SET faviconID = ? WHERE bmkUri = ?", withArgs: [id, site.url])
+                    try? conn.executeChange("UPDATE bookmarksMirror SET faviconID = ? WHERE bmkUri = ?", withArgs: [id, site.url])
+
+                    return id
+                }
+
+                let err = DatabaseError(description: "Error adding favicon. ID = 0")
+                log.error("addFavicon(_:, forSite:) encountered an error: \(err.localizedDescription)")
+                throw err
+            }
+        }
+
+        let siteSubselect = "(SELECT id FROM history WHERE url = ?)"
+        let iconSubselect = "(SELECT id FROM favicons WHERE url = ?)"
+        let insertOrIgnore = "INSERT OR IGNORE INTO favicon_sites (siteID, faviconID) VALUES "
+        if let iconID = icon.id {
+            // Easy!
+            if let siteID = site.id {
+                // So easy!
+                let args: Args? = [siteID, iconID]
+                return doChange("\(insertOrIgnore) (?, ?)", args: args)
+            }
+
+            // Nearly easy.
+            let args: Args? = [site.url, iconID]
+            return doChange("\(insertOrIgnore) (\(siteSubselect), ?)", args: args)
+
+        }
+
+        // Sigh.
+        if let siteID = site.id {
+            let args: Args? = [siteID, icon.url]
+            return doChange("\(insertOrIgnore) (?, \(iconSubselect))", args: args)
+        }
+
+        // The worst.
+        let args: Args? = [site.url, icon.url]
+        return doChange("\(insertOrIgnore) (\(siteSubselect), \(iconSubselect))", args: args)
+    }
+
+    public func getFaviconImage(forSite site: Site) -> Deferred<Maybe<UIImage>> {
+        // First, attempt to lookup the favicon from our bundled top sites.
+        return getTopSitesFaviconImage(forSite: site).bind { result in
+            guard let image = result.successValue else {
+
+                // Second, attempt to lookup the favicon URL from the database.
+                return self.lookupFaviconURLFromDatabase(forSite: site).bind { result in
+                    guard let faviconURL = result.successValue else {
+
+                        // If it isn't in the database, attempt to scrape its
+                        // URL from the web page.
```

Comment:
nit: remove comments please, would be easier to read these nested-if blocks when more compacted.

Reply:
Really? I wanted the comments here to explain clearly to the reader in the future exactly what the process is.

Reply:
Maybe one-line the comments and remove some of the newlines?

Reply:
Ok, yeah that would be better. This is such a convoluted process of falling back to alternative icons sources that I want it to be crystal clear what is going on the next time someone needs to come in here.